File: src/0927/S0927_v1.py
class Solution:
    def threeEqualParts(self, arr):
        cnt = arr.count(1)
        if cnt == 0:
            return [0, len(arr)-1] # 全是0时i, j取头尾
        if cnt % 3 != 0:
            return [-1, -1] # 1的数量不能被3整除即不可能被3等分
        j = cnt // 3
        cnt, i = 0, 0
        for v in arr[::-1]:
            i += 1
            if v == 1:
                cnt += 1
                if cnt == j:
                    j = len(arr) - i
                    break
        cnt = 0
        while cnt < j:
            if arr[cnt] == 1:
                break
            cnt += 1
        r = len(arr[j:])
        i = cnt + r - 1
        if arr[cnt:i+1] != arr[j:]:
            return [-1, -1]
        cnt = i + 1
        while cnt < j:
            if arr[cnt] == 1:
                break
            cnt += 1
        if arr[cnt:cnt+r] != arr[j:]:
            return [-1, -1]
        return [i, cnt+r]

File: src/0927/S0927_v1.rs
impl Solution {
    pub fn three_equal_parts(arr: Vec<i32>) -> Vec<i32> {
        let alen = arr.len();
        let mut cnt = 0;
        let mut a1 = 0;
        for v in (0..alen) {
            if arr[v] == 1 {cnt += 1;a1 = v; break;}
        }
        for v in (a1+1..alen) {
            if arr[v] == 1 {cnt += 1;}
        }
        if cnt == 0 {return [0, alen as i32 - 1].to_vec()};
        let fail = vec![-1, -1];
        if cnt % 3 != 0 {return fail};
        for v in (0..alen) {
            if arr[v] == 1 {a1 = v; break;}
        }
        // let a1 = arr.iter().take_while(|&&x| x!=1).count();
        let mut a2 = 2 * cnt as usize / 3;
        let mut a3 = cnt as usize / 3;
        let mut cnt = 0 as usize;
        for v in (a1..alen).rev() {
            if arr[v] == 1 {
                cnt += 1;
                if (cnt == a3 as usize) {
                    a3 = v;
                    break;
                }
            }
        }
        let r = alen - a3;
        for v in (a1+r..a3) {
            if arr[v] == 1 {
                a2 = v;
                break;
            }
        }
        if a1+r > a2 || a2+r > a3 {return fail;}
        let mut i: usize = a1;
        let mut j: usize = a2;
        for v in (a3..alen) {
            if arr[v] != arr[i] || arr[v] != arr[j] {return fail;}
            i += 1;
            j += 1;
        }
        [i as i32 - 1, j as i32].to_vec()
    }
}
